src/solvers/flattening/boolbv_array.cpp

```diff
@@ -11,17 +11,16 @@ Author: Daniel Kroening, kroening@kroening.com
 
 bvt boolbvt::convert_array(const exprt &expr)
 {
-  std::size_t width=boolbv_width(expr.type());
+  const std::size_t width = boolbv_width(expr.type());
+  const exprt::operandst &operands = expr.operands();
 
-  if(width==0)
-    return conversion_failed(expr);
+  if(operands.empty() && width == 0)
```

Comment:
Is this adding functionality or is this just suppressing an error case?  What is the use-case for this?

Reply:
This is removing a warning in case there is an equality between empty arrays (this was occurring in some of the tests, when I added the java models library to the classpath)

Reply:
OK but why is the library model comparing empty arrays?  Maybe there is a good reason but equally, it could be a bug, right?

Reply:
I think it's because there are some empty string literals there. I not sure about the exact reason, but I don't see anything wrong with comparing empty arrays, no?

Reply:
Is an empty string actually an empty array? I'm with Martin, this seems worth investigating. Maybe there are good reasons, but those reasons haven't come up in the last 15 years and thus should be free from doubt and guesswork.

Reply:
@romainbrenguier : I doubt it is intrinsically harmful but I suspect that quite a bit of the rest of the array handling code was written with the implicit assumption that the arrays were non-empty.  Also, whenever someone removes defensive code in favour of handling a larger set of programs, I like to make sure it isn't a case of "the fire alarms were annoying so we turned them off".  This could quite well be the right thing to do; I'd just like to know why.

Reply:
Maybe we should ask the question differently: 'What is the error case that above conditional is supposed to detect?'

Reply:
`boolbv_width` would return `0` for an array of non-constant size. That, however, seems like a bogus case here as a concrete array for a type of unknown size is (at least to me) hard to imagine. So it's probably true that the only reasonable case is that of zero operands. Would any call site of this function fail when receiving an empty array? It's hard to tell, but as of 70532f992716 we are in a much better shape to detect those.